Automattic/woocommerce-payments
# refactor: hooks and selector dependencies

#### Changes proposed in this Pull Request









The way the hooks are currently written creates a performance degradation in the way the page is rendered when an action is taken by the merchant.
So, if  a checkbox is interacted with (as an example), the whole page re-renders, instead of re-rendering only the area surrounded by the checkbox where the hook is called.

The issue is evident in these captures:
Before - most sections are re-rendered:
![2023-10-27 10 36 06](https://github.com/Automattic/woocommerce-payments/assets/273592/4d83eb6b-e5d7-494b-917c-849aab6a1f35)

After - only the needed sections are re-rendered:
![2023-10-27 11 27 19](https://github.com/Automattic/woocommerce-payments/assets/273592/ab071b7e-9d2a-430f-af95-abf72f87e3b9)


Refactoring, so that the hooks are no longer written like this:
```js
const useSomeSetting = () => {
	const { updateSomeSetting } = useDispatch( STORE_NAME );

	return useSelect(
		( select ) => {
			const { getSomeSetting } = select( STORE_NAME );

			return [
				getSomeSetting(),
				updateSomeSetting,
			];
		},
		[ updateSomeSetting ]
	);
};
```

But instead it is written like this:
```js
const useSomeSetting = () => {
	const { updateSomeSetting } = useDispatch( STORE_NAME );

	const someSetting = useSelect(
		( select ) => {
			const { getSomeSetting } = select( STORE_NAME );

			return getSomeSetting();
		}
	);

  return [ someSetting, updateSomeSetting ];
};
```

The difference is that the `updateSomeSetting` action is no longer a dependency of the selector.

#### Testing instructions





- Go to the settings page(s)
- Ensure that changing/saving the settings still works

If you'd like to test the re-rendering times:
- Go to the settings page
- Open the Dev Tools
- Ensure you have the React Dev Tools installed
- Go to the "Components" tab
- Enable the "Highlight updates when components render" option
![Markup 2021-07-14 at 15 32 54](https://user-images.githubusercontent.com/273592/125689092-1f9b1645-abc6-46ba-aae3-f5885148b117.png)

-------------------


**Post merge**
